### DISCUSSION TITLE
Null-conditional yield return

### DISCUSSION THREAD

**Discussion by Rekkonnect** [dotnet/csharplang#5063] [Language Ideas]
# Summary

New syntax sugar to remove the check for whether the returned value is `null` or not.

# Motivation

I've encountered cases where I want to `yield return` if a value is not null, which in my opinion can become a hussle to deal with, especially if the returning value is yielded from an expensive operation.

# Example

```csharp
var expensiveValue = ExpensiveOperation();
if (expensiveValue != null)
    yield return expensiveValue;
```
could become something like
```csharp
yield return? ExpensiveOperation();
```

# Drawbacks

Since `?` indicates nullity in several contexts, it could be slightly confusing to the user as to why it is allowed there and why it prevents returning if `null`. However, `!` cannot be used because it is the unary logical negation operator. Tying it right next to the `return` keyword would also not look that great.

Another solution could be using the `notnull` contextual keyword. In that case, the syntax would be
```csharp
yield return notnull value;
```

Its main issue is that then the expression becomes too verbose.

**Comment by YairHalberstadt** [comment]
What's wrong with the current way of doing this? Why is this case so important that special syntax is needed for it?

**Reply by Rekkonnect** [reply]
It's not wrong, it's just an improvement like the `?.` operator. All it does is elimiate at least one if statement, if applicable, and maybe a redundant variable declaration.

**Reply by YairHalberstadt** [reply]
Note you can currently do:

```csharp
if (ExpensiveOperation is {} val) yield return val;
```

**Reply by YairHalberstadt** [reply]
The `?.` operator is useful because it can often remove many many statements, and because it's used all the time.

This would remove at most one statement and is fairly niche in it's usage.

The cost is that every time syntax is introduced, people have to learn that syntax. It's much easier to read if statements which we all know and recognize, than some new syntax one off just for this case. The `?.` operator is actually quite confusing, and a bit of a hurdle for new developers, but so useful that it's worth the cost anyway.

**Reply by Rekkonnect** [reply]
Sure thing, but that's just another (shorter) way to declare the variable and check if it's not null. You don't avoid the whole thing.

**Reply by WhiteBlackGoose** [reply]
It's not a bad idea. But it shouldn't be hardcoded for this one case. It could be instead extended to F#'s computation expressions, and `yield return? (T?)` being a BCL thing instead of a language thing.

**Comment by Abrynos** [comment]
The idea of syntax sugar for that is not bad in my opinion, but I'd prefer if it was something like this:
```csharp
object? value = GetValueOrNull();

yield return value ?? continue;// return or continue the iteration
yield return value ?? break;// return or end the iteration
yield return value ?? throw new SomeException(42);// return or throw exception - already possible
```

**Reply by Rekkonnect** [reply]
That's a generalizable suggestion that could also apply to `return`.

As for how the proposal could be implemented, `yield return value ?? continue;` is lengthier and I wouldn't favor it for that reason.

**Reply by FaustVX** [reply]
The long length of a syntax is not a valid reason to invalidate a proposal.
The readiness & comprehension is important.
So, the `??` syntax is a good candidate, and also, it's flexible.

**Reply by Abrynos** [reply]
> That's a generalizable suggestion that could also apply to `return`.

Correct. The `throw`-part is already possible, but I think the `continue` or `break` one would be a nice addition.

> `yield return value ?? continue;` is lengthier

Is it?
```csharp
while(SomeCondition())
{
    object? value = GetValueOrNull();
    // ... do something assigning/further computation on value
    yield return value ?? continue;
    // ... some more logic
}
```
```csharp
while(SomeCondition())
{
    object? value = GetValueOrNull();
    // ... do something assigning/further computation on value
    if(value != null)
    {
        yield return value;
    }
    // ... some more logic
}
```

**Reply by Rekkonnect** [reply]
The part about your proposal being lengthier refers to that when compared to my `yield return? value`, which is a more specific case. In other words, it could be considered shorthand for `yield return value ?? continue;`. Or better yet, since `continue` is not a valid keyword in non-loop context, it would be wiser to not further confuse the user by requiring such usage and adjusting its behavior depending on the scope.

What I mean is, `yield return value ?? continue;` is nice, when you can actually replace
```csharp
if (value is null)
    continue;
yield return value;
```
and not just continue with the rest of the code. Similar is the case with `break`.

All in all, `yield return? value` has its own place because there is no expression to indicate that code execution should proceed normally. Therefore, all those new cases are handy additions that I can't argue against easily, but aren't bound to this proposal, although related.

**Comment by acaly** [comment]
I think it will be made more useful if extended to other expressions. Currently `?.` only works if the checked expression is used before a `.`, not elsewhere. Maybe we can introduce an ignore or skip keyword to be used with `??`, for example:

```c#
obj.M(anotherObj ?? skip);
(_event ?? skip)(this, args);
yield return val ?? skip;
```

Evaluation order will be a problem, I think.

**Reply by Rekkonnect** [reply]
As per your examples,
```csharp
_event?(this, args);
```
would do a better job.

`?? skip` seems too junky of a keyword-ish. I mean, even `yield return?` kinda does, for the same reason `delete[]` in C++ looks like so.

```csharp
if (anotherObj is not null)
    obj.M(anotherObj);
```
is too special a case, and it can quickly get awful if it becomes an available solution for multiple arguments at the same time. Imagine
```csharp
obj.M(a ?? skip, b ?? skip, c ?? skip);
```

**Reply by acaly** [reply]
I know it's personal preference, but I would like a slightly longer syntax, as it's really something that shouldn't be overlooked. A single ? in  `func?()` might not catch people's attention enough.
